### DISCUSSION TITLE
Test location provider in Out-JSONFile

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [SamErde/PSPreworkout#14] [General]
Since `Out-JSONFile` is writing a file to the current location, it would be nice if you could validate that the location is a Filesystem location before trying to write the file.

![image](https://github.com/user-attachments/assets/008816d9-e54d-4b89-b9b5-fbb8aa8743fa)

**Comment by SamErde** [comment]
Thanks, I like that idea! This function was a pretty quick creation and I'm sure there will be good ways to further solidify it.

**Reply by SamErde** [reply]
I just realized that relying on $PWD as the default path parameter value is what causes this. Time for a different default (if any).

**Comment by SamErde** [comment]
@jdhitsolutions, I tried a few different approaches and ended up with this commit https://github.com/SamErde/PSPreworkout/blob/552592b2dc7aa8cca3980c3dd32f945a8cf3acde/src/PSPreworkout/Public/Out-JsonFile.ps1. I *think* this covers all of your suggestions. 🙏

**Comment by jdhitsolutions** [comment]
You could move the test for the json extension to a validate pattern attribute.

```powershell
[ValidatePattern('\.json$')]
```

You could save yourself a lot of work and simply save the JSON file to `$ENV:Temp` by default.

**Reply by SamErde** [reply]
Oh, that ValidatePattern is brilliantly simple!

**Reply by SamErde** [reply]
A colleague suggested that I not assume a required .json extension because of the many use cases where a different extension may be required. 🧐 It is a good point. I believe the extension can be a good indicator of the contents of the file and the op can always rename it, _but_ the argument does still make me want to adjust my strategy: use a .json extension on the default filename if none is specified but otherwise allow the op to specify a file with any/no extension if they want.

**Reply by jdhitsolutions** [reply]
Then I would at least use `[ValidateNotNullOrEmpty()]` on the file path parameter. Don't assume a user is going to type something. You have no idea on where the parameter value is coming from. They may be using a variable defined somewhere else, or at least they think it was defined. I think many scripters assume someone is going interactively run a command and type parameter values. You have no idea on how someone might use your command and how they will pass parameter values.

**Comment by SamErde** [comment]
That is easy enough to implement, but it does bring me back to one of the paths that I was debating earlier: what is the best practice for where to perform validation, and why? I can seem some pros and cons to using parameter validation, and some for performing validation with handling in the begin block.

For example, I could add the following ValidateScript to check whether the directory actually exists or not...but if I take this approach to its full extent, then half of the script now exists in the param () block. 

```powershell
function Test {
    [CmdletBinding()]
    param (

        # Full path and filename to save the JSON to.
        [Parameter(Position = 1)]
        [ValidatePattern('\.json$')]
        [ValidateScript({
                if ((Split-Path -Path $_).Length -gt 0) {
                    if (Test-Path -Path (Split-Path -Path $_) -PathType Container) {
                        $true
                    } else {
                        throw "The directory `'$(Split-Path -Path $_)`' was not found."
                    }
                } else { $true }
            })]
        [string]
        $FilePath
    )

    Write-Host $FilePath
}

Test 'test.json'
Test "$env:Temp\test.json"
Test
```

Good / bad / it depends?

**Reply by jdhitsolutions** [reply]
Put the code where it needs to be.

**Comment by SamErde** [comment]
Does it come down to "what do I want the UX to be like?" (should it terminate with an invalid or incomplete argument, or should I try to anticipate potential issues and work around those in the body?) or would you bring up other issues for consideration?

**Comment by jdhitsolutions** [comment]
I am an old school IT Pro. When I run a command, I want to know it will run successfully. If there is going to be a problem, like a bad file name, I want to know up front before anything can happen.  You don't want to force the user to jump through hoops, but you should ensure that the command has everything it needs to execute successfully.
